roxctl/README.md

```diff
@@ -0,0 +1,189 @@
+# Introduction
+
+`roxctl` is the CLI for Stackrox.
+
+It currently serves as a multi-purpose CLI that handles things such as:
+- administrative tasks (e.g. creating backups / restoring them)
+- integration within CI for image / workload scanning.
+- acting as API client for automation purposes.
+
+The goal of this README is to provide an overview of writing and maintaining commands of `roxctl`.
+
+# Command structure
+
+Commands **must** follow the common name convention: `roxctl <noun> [subcommands...]`.
+For more information on why this decision was made, [an ADR exists explaining them](https://github.com/stackrox/architecture-decision-records/blob/main/stackrox/ADR-0004-roxctl-subcommands-layout.md).
+
+Within the codebase, `roxctl` makes heavy use of [github.com/spf13/cobra](https://pkg.go.dev/github.com/spf13/cobra) as
+the CLI library of choice.
+
+Besides `github.com/spf13/cobra`, one integral part how commands are structured is the [roxctl/common/environment](https://github.com/stackrox/stackrox/tree/master/roxctl/common/environment) package.
+The `Environment` interface provides abstractions for generic functionality that may be required by a command:
+- Printing output.
+- Creating gRPC connections / HTTP client.
+- Logging.
+- Input / Output abstractions.
+- Config store.
+
+Each command used in `roxctl` **must** use the `Environment` interface. It provides a clear structure to all commands,
+as well as the possibility to easily mock external dependencies to allow for better and sophisticated unit testing.
+
+A typical command structure within Go should look as follows:
+```go
+package sample
+
+// Command exposes the *cobra.Command for your specific command. It will **always** take a environment.Environment
+// as input.
+func Command(cliEnvironment environment.Environment) *cobra.Command {
+    sampleCmd := &sampleCmd{env: cliEnvironment}
+
+	cmd := &cobra.Command{
+		Use: "sample",
+		Short: "A sample command",
+		Long: `A sample command.
+This is a sample for a multi-line detailed description for the sample command`,
+        Args: cobra.NoArgs,
+		// Each command **must** always use `RunE` and return an error.
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if err := sampleCmd.Construct(cmd); err != nil {
+				return err
+			}
+			return sampleCmd.sample()
+		},
+    }
+
+	// Each custom flag for a command **must** be bound to the struct holding all relevant properties for a command.
+	cmd.Flags().StringVarP(&sampleCmd.name, "name", "n", "", "name to say hello to")
+
+	return cmd
+}
+
+// Sample command holds the cli environment, as well as all resolved flag values. This makes unit testing easier.
+type sampleCmd struct {
+	env environment.Environment
+
+	name string
+}
+
+// Construct ensure all flag values are resolved and possibly validated. Any additional external dependency that you may
+// require (e.g. a gRPC client) may be constructed here as well.
+func (s *sampleCmd) Construct(cmd *cobra.Command) error {
```

Comment:
When adding the environment we thought about it, but didn't do it back then.

It might be worthwhile to require this as an interface in the future, we can enhance it afterwards. I'd like to make the code changes before adding it to the README though.

Reply:
Sounds good, thank you!